test/testutil.h

```diff
@@ -141,10 +143,201 @@ __owur int run_tests(const char *test_prog_name);
+    DECLARE_COMPARISON(type, name, opname)                              \
+    DEFINE_COMPARISON(type, name, opname, op, fmt)
+
+#define IMPLEMENT_COMPARISONS(type, name, fmt)                          \
+    DECLARE_COMPARISONS(type, name)                                     \
+    DEFINE_COMPARISONS(type, name, fmt)
+
+/*
+ * Define some comparisons between pairs of various types.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ *
+ * In each case the functions produced are:
+ *  int test_name_eq(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_ne(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_lt(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_le(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_gt(const type t1, const type t2, const char *desc, ...);
+ *  int test_name_ge(const type t1, const type t2, const char *desc, ...);
+ *
+ * The t1 and t2 arguments are to be compared for equality, inequality,
+ * less than, less than or equal to, greater than and greater than or
+ * equal to respectively.  If the specified condition holds, the functions
+ * return 1.  If the condition does not hold, the functions print a diagnostic
+ * message and return 0.
+ *
+ * The desc argument is a printf format string followed by its arguments and
+ * this is included in the output if the condition being tested for is false.
  */
-int strings_equal(const char *desc, const char *s1, const char *s2);
+DECLARE_COMPARISONS(int, int)
+DECLARE_COMPARISONS(unsigned int, uint)
+DECLARE_COMPARISONS(char, char)
+DECLARE_COMPARISONS(unsigned char, uchar)
+DECLARE_COMPARISONS(long int, lint)
+DECLARE_COMPARISONS(unsigned long int, ulint)
+
+/*
+ * Pointer comparisons against other pointers and null.
+ * These functions return 1 if the test is true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(void *, ptr, eq)
+DECLARE_COMPARISON(void *, ptr, ne)
+int test_ptr(const void *, const char *, ...);
+int test_ptr_null(const void *, const char *, ...);
+
+/*
+ * Equality tests for strings where NULL is a legitimate value.
+ * These calls return 1 if the two passed strings compare true.
+ * Otherwise, they return 0 and pretty-print diagnostics.
+ */
+DECLARE_COMPARISON(char *, str, eq)
+DECLARE_COMPARISON(char *, str, ne)
+
+/*
+ * Check a boolean result for being true or false.
+ * They return 1 if the condition is true (i.e. the value is non-zro).
+ * Otherwise, they return 0 and pretty-prints diagnostics using |desc|.
+ */
+int test_true(int, const char *desc, ...);
+int test_false(int, const char *desc, ...);
+
+/*
+ * The following macros provide wrapper calls to the test functions with
+ * a default description that indicates the file and line number of the error.
+ */
+#define TEST_int_eq(a, b)   test_int_eq(a, b, "%s: %d", __FILE__, __LINE__)
```

Comment:
This looks good! However I'd still like to preserve file and line numbers for the custom error messages too. Now we can't do that in general (because no support for variadic macros in C90) but you could do something like this I think?

```
# if !defined(__STDC_VERSION__) || __STDC_VERSION__ < 199901L
# define TEST_error test_error_c90
#else
# define TEST_error(...) test_error(__FILE__, __LINE__, __VA_ARGS__)
#endif
```

where `test_error_c90` takes a format string and args, and `test_error` additionally takes the file and line number.

Then hide the lowercase functions in a separate internal header (`testutil_internal.h`) so that tests don't use them directly. Code that previously called `test_int_eq` would do

```
if (!TEST_int_eq(a[i], b[i])) {
  TEST_error("%d th element mismatch", i);
}
```

Or you could also provide a macro per comparison, `TEST_int_eq_msg(a, b, msg)`, etc.

What do you think?